tests/unit/ir/schinks.ml

```diff
@@ -0,0 +1,406 @@
+open Links_core
+open Utility
+
+(* look at the .mli file if you just want to use this to write test cases *)
+
+module CT = CommonTypes
+
+module Repr = struct
+  type name_maps = {
+    tname_to_id : int StringMap.t;
+    varname_to_id : int StringMap.t;
+  }
+
+  (* [@@deriving show] *)
+
+  type 'a maker = {
+    mk_ir_form : name_maps -> 'a;
+    (* [@printer fun fmt _ -> ()] *)
+    reserved_tids : IntSet.t;
+    reserved_ids : IntSet.t;
+    used_names : string list;
+    used_tnames : string list;
+  }
+
+  (* [@@deriving show] *)
+
+  type 'a t = ('a maker, string) Result.t
+
+  let make (maker : 'a maker) : 'a =
+    let rec add_binding reserved next_id used map =
+      match (used, IntSet.mem next_id reserved) with
+      | [], _ -> map
+      | _, true -> add_binding reserved (next_id + 1) used map
+      | name :: rem_names, false ->
+          add_binding reserved (next_id + 1) rem_names
+            (StringMap.add name next_id map)
+    in
+    let varname_to_id =
+      add_binding maker.reserved_ids 0 maker.used_names StringMap.empty
+    in
+    let tname_to_id =
+      add_binding maker.reserved_tids 0 maker.used_tnames StringMap.empty
+    in
+    maker.mk_ir_form { varname_to_id; tname_to_id }
+
+  (* Note that this isn' quite monadic bind due to the param type of the passed function *)
```

Comment:
This seems rather strange.  I strongly suspect that the `([-] maker, string) Result.t` is really a well-behaved monad in disguise, probably a combination of Reader/Writer and exception monads.  (Possibly also an applicative/monoidal functor as well.)

Concretely, an `'a t` is something which:
- can fail, returning a string
- if not failing, returns:
  - a function from name_maps to 'a
  - other stuff which was collected up during construction of 'a (`reserved_*ids`, `used_*names`)

I think it does no harm to hoist the name_maps, to give type
```
type 'a maker = name_maps -> {ir_form: 'a, reserved_tids: ...} Result.t
```
and then return would be the obvious thing (ignore the name_maps, generate empty maps/lists), and bind would just be the composition of the exception, writer and reader monad binds.

I'm not sure how much the rest of the code depends on the non-standard bind though - my hope would be that using a more standard type and setup would simplify matters but YMMV.

Reply:
Thanks, that's a good point. I suspected that there's a proper monad in there, but didn't quite manage to find it. @rudihorn expressed some interest in working on this, otherwise I will refactor it myself at the end of the week.